pkg/gce-pd-csi-driver/controller.go

```diff
@@ -431,14 +432,118 @@ func (gceCS *GCEControllerServer) ControllerGetCapabilities(ctx context.Context,
+	}
+	if len(req.Name) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "CreateSnapshot Snapshot name must be provided")
+	}
+	snapshot, err := gceCS.CloudProvider.CreateSnapshot(ctx, volKey, req.Name)
+	if err != nil {
+		return nil, status.Error(codes.Internal, fmt.Sprintf("Unknown create snapshot error: %v", err))
+	}
+	t, err := time.Parse(time.RFC3339, snapshot.CreationTimestamp)
+	if err != nil {
+		return nil, status.Error(codes.Internal, fmt.Sprintf("Failed to covert creation timestamp: %v", err))
+	}
+	createResp := &csi.CreateSnapshotResponse{
+		Snapshot: &csi.Snapshot{
+			SizeBytes:      	common.GbToBytes(snapshot.DiskSizeGb),
+			Id:                 cleanSelfLink(snapshot.SelfLink),
+			SourceVolumeId:     volumeID,
+			CreatedAt: 			t.UnixNano(),
+			Status: &csi.SnapshotStatus{
+				Type: convertCSISnapshotStatus(snapshot.Status),
+			},
+		},
+	}
+	return createResp, nil
+}
+
+func convertCSISnapshotStatus(status string) csi.SnapshotStatus_Type {
+	var csiStatus csi.SnapshotStatus_Type
+	switch status {
+	case "READY":
+		csiStatus = csi.SnapshotStatus_READY
+	case "UPLOADING":
+		csiStatus = csi.SnapshotStatus_UPLOADING
+	case "FAILED":
+		csiStatus = csi.SnapshotStatus_ERROR_UPLOADING
+	case "DELETING":
+		csiStatus = csi.SnapshotStatus_UNKNOWN
+		glog.V(4).Infof("snapshot is in DELETING")
+	default:
+		csiStatus = csi.SnapshotStatus_UNKNOWN
+	}
+	return csiStatus
 }
 
 func (gceCS *GCEControllerServer) DeleteSnapshot(ctx context.Context, req *csi.DeleteSnapshotRequest) (*csi.DeleteSnapshotResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "")
+	glog.V(4).Infof("DeleteSnapshot called with request %v", *req)
+
+	// Validate arguments
+	snapshotID := req.GetSnapshotId()
+	if len(snapshotID) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "DeleteSnapshot Snapshot ID must be provided")
+	}
+
+	key, err := common.SnapshotIDToKey(snapshotID)
+	if err != nil {
+		// Cannot get snapshot ID from the passing request
+		return nil, status.Error(codes.InvalidArgument, "snapshot id is not right")
+	}
+
+	err = gceCS.CloudProvider.DeleteSnapshot(ctx, key)
+	if err != nil {
+		return nil, status.Error(codes.Internal, fmt.Sprintf("unknown Delete snapshot error: %v", err))
+	}
+
+	return &csi.DeleteSnapshotResponse{}, nil
 }
 
 func (gceCS *GCEControllerServer) ListSnapshots(ctx context.Context, req *csi.ListSnapshotsRequest) (*csi.ListSnapshotsResponse, error) {
+	glog.V(4).Infof("ListSnapshots called with request %v", *req)
+
+	// In the first version, only support ListSnapshots if snapshot id is specified.
+	// If no snapshot id is specified, return Unimplemented error code
+	if snapshotId := req.GetSnapshotId(); snapshotId != "" {
+		key, err := common.SnapshotIDToKey(snapshotId)
+		if err != nil {
+			// Cannot get snapshot ID from the passing request
+			return nil, status.Error(codes.InvalidArgument, "snapshot id is not right")
```

Comment:
nit: add debug string such as the snapshotId provided

Reply:
done